src/main/java/games/powergrid/PowerGridGameState.java

```diff
@@ -0,0 +1,941 @@
+	    // First, check current market
+	    if (currentMarket != null) {
+	        for (PowerGridCard card : currentMarket.getComponents()) {
+	            if (card.getNumber() == cardNumber) {
+	                return false;
+	            }
+	        }
+	    }
+
+	    // Then, check future market
+	    if (futureMarket != null) {
+	        for (PowerGridCard card : futureMarket.getComponents()) {
+	            if (card.getNumber() == cardNumber) {
+	                return true;
+	            }
+	        }
+	    }
+
+	    // If not found in either, return null
+	    return false;
+	}
+    public int getPlayerCapacity(int playerId) {
+    	int capacity = 0; 
+    	for (PowerGridCard card : this.getPlayerPlantDeck(playerId)) {
+    		capacity += card.getCapacity(); 
+    	}
+    	return capacity; 
+    }
+    
+    public int [] getPlayerCapacity() {
+    	int[] playersCapacity = new int[this.getNPlayers()];
+    	for(int i = 0; i < this.getNPlayers(); i++) {
+    		playersCapacity[i] = getPlayerCapacity(i);
+    	}
+    	return playersCapacity;
+    }
+    
+    public int totalMoney() {
+    	return Arrays.stream(playerMoney).sum();  	
+    }
+    
+    public void setOneHotRegions(Set<Integer> activeRegions, int numRegions) {
+        double[] out = new double[numRegions];
+        for (int i = 1; i <= numRegions; i++) {
+            out[i - 1] = activeRegions.contains(i) ? 1 : 0;
+        }
+        this.oneHotRegion = out;
+    }
+    
+    public double[] getOneHotRegions() {
+    	return this.oneHotRegion;
+    }
+    
+    /**
+     * Computes the game score for the given player, used for intermediate
+     * reward shaping and final outcome scoring. This is what PyTAG uses so this does not 
+     * necessarily reflect the "Score" of a game. 
+     *
+     * <p>After the Bureaucracy phase, the score is calculated based on a 
+     * a weighted combination of:
+     * <ul>
+     *   <li>Normalized city count (progress toward game-end city target)</li>
+     *   <li>Normalized income (based on powered cities)</li>
+     * </ul>
+     *
+     * <p>At game end:
+     * <ul>
+     *   <li>Winners receive +4.0</li>
+     *   <li>All non-winners receive −2.0</li>
+     * </ul>
+     *
+     * @param playerId the player whose score is being calculated
+     * @return a reward value reflecting current progress or final result
+     */
+
+    @Override
+    public double getGameScore(int playerId) {
+        final PowerGridGamePhase phase = (PowerGridGamePhase) getGamePhase();
+        
+        final double wCities = 0.8;
+        final double wIncome = 1.2;
+        
+ 
+       //if the player wins at the end they get a reward else they get penalized 
+        if (this.getGameStatus() == CoreConstants.GameResult.GAME_END) {
+            CoreConstants.GameResult[] res = getPlayerResults();
+            CoreConstants.GameResult r = (res != null && playerId < res.length) ? res[playerId] : null;
```

Comment:
The game score should always be the actual game score, there shoudl be no need to check for GAME_END. For Powergrid I would argue that the score is the number of powered citie; plus tiebreak of money.

(This differs to getHeuristicScore which can/should check for game end and normalise/clamp to a [-1, +1] range; and can use any number of heuristic estimates